lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -0,0 +1,312 @@
+//===- RefineTypes.cpp ------------------------*- C++-*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "npcomp/Dialect/ATen/IR/ATenDialect.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyDialect.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyOps.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyDialect.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyOps.h"
+#include "npcomp/Dialect/Torch/IR/TorchDialect.h"
+#include "npcomp/Dialect/Torch/IR/TorchOps.h"
+#include "npcomp/Dialect/Torch/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::NPCOMP;
+using namespace mlir::NPCOMP::Torch;
+
+// -----------------------------------------------------------------------------
+// Analysis.
+// -----------------------------------------------------------------------------
+
+constexpr int64_t kUnknownSize = -1;
+
+namespace {
+// Statically known information for a particular Value.
+//
+// This struct currently tracks only information relevant for tensor/array-like
+// shaped types. It is fine to associate a `ValueKnowledge` with a non-shaped
+// type as long as it is in the default "no knowledge" state returned by
+// `getMostConservativeKnowledge`. The important invariant is that we cannot
+// claim to know something about a value which is false.
+//
+// This class could also be called "dataflow facts", "lattice value", etc.
+struct ValueKnowledge {
+  ValueKnowledge() = delete;
+  // We enforce that `elementType` is always a valid type (possibly
+  // !numpy.any_dtype), and `sizes` is empty unless `hasRank`.
+  // So default constructing is prohibited.
+  ValueKnowledge(bool hasRank, std::vector<int64_t> sizes, Type elementType)
+      : hasRank(hasRank), sizes(sizes), elementType(elementType) {
+    assert(elementType != nullptr);
+    assert(sizes.size() == 0 || hasRank);
+  }
+
+  // Get a safe "most conservative knowledge" default state.
+  static ValueKnowledge getMostConservativeKnowledge(MLIRContext *context) {
+    return ValueKnowledge(false, {}, Numpy::AnyDtypeType::get(context));
+  }
+
+  // Whether the Value is known to have a rank.
+  bool hasRank;
+  // If `hasRank` the sizes along each rank. Unknown sizes are represented as
+  // `kUnknownSize`.
+  std::vector<int64_t> sizes;
```

Comment:
I haven't really profiled/optimized this. And as a matter of convention, generally SmallVector isn't the default "go to" on the heap (it can blow up RSS quickly).